Remove implicit conversions between points and weighted points

MaelRL opened the pull request:
## Summary of Changes
The current implementation of weighted points in the CGAL kernel allows for implicit conversions between points and weighted points, meaning that the compiler can convert a `Weighted_point_23` into a `Point_23` (and reversely) if it allows to match a function. This is dangerous because we can accidentally drop weights or call the wrong function unintentionally.

This small feature does not remove these conversions but makes them explicit, meaning that the constructor must be called explicitly to be used (and thus the compiler is not allowed to use them on its own). 

## Release Management

* Affected package(s): Main changes in kernel packages, incident changes  a bit everywhere (mostly `Alpha_shapes_23`, `Interpolation`, `Kinetic_data_structures`, `Mesh_3`, `Skin_surface_3`, and `Triangulation_23`). 
* Issue(s) solved (if any): fix #2144, fix #1878
* Feature/Small Feature (if any): [Remove Point -- Weighted point implicit conversions](https://cgal.geometryfactory.com/CGAL/Members/wiki/Features/Small_Features/Remove_Point_--_Weighted_point_implicit_conversions)

janetournois commented:
Here is the list of folders in which changes need to be reviewed 
- [ ] AABB_tree
- [ ] Alpha_shapes_2
- [ ] Alpha_shapes_3
- [ ] Apollonius_graph_2
- [ ] BGL
- [ ] Cartesian_kernel
- [ ] CGAL_ipelets
- [ ] Generator
- [ ] GraphicsView
- [ ] Homogeneous_kernel
- [ ] Installation
- [ ] Interpolation
- [ ] Kernel_23
- [ ] Kinetic_data_structures
- [ ] Kinetic_framework
- [ ] Mesh_3
- [ ] Partition_2
- [ ] Periodic_triangulation_2
- [ ] Periodic_triangulation_3
- [ ] Poisson_surface_reconstruction_3
- [ ] Polyhedron
- [ ] Property_map
- [ ] Segment_Delaunay_graph_2
- [ ] Skin_surface_3
- [ ] STL_Extension
- [ ] Surface_mesher
- [ ] Triangulation_2
- [ ] Triangulation_3
- [ ] Voronoi_diagram_2

sloriot commented on Alpha_shapes_3/doc/Alpha_shapes_3/CGAL/Alpha_shape_3.h:
```diff
@@ -7,21 +7,33 @@ namespace CGAL {
+\warning The tag `ExactAlphaComparisonTag` is currently ignored (that is, the code will
+ behave as if `ExactAlphaComparisonTag` were set to \link Tag_false `Tag_false`\endlink)
+if the traits defines a point type that is not implicitely convertible to the three-dimensional point type
+of a CGAL kernel. This is because we internally use the class Cartesian_converter to switch
```
backticks missing

sloriot replied:
An alternative here is to document that `Tag_true` can be used only if `Cartesian_converter` can convert a Point/Weighted Point from the traits to any CGAL kernel.

MaelRL replied:
I've kept the warning in negative form, but simplified it a bit and regrouped it with the remark about using this tag in conjonction with an exact kernel, see 84437a4.

sloriot replied:
Maybe I wasn't clear. What I meant is that I would remove the `boost::is_convertible` and document that for example `Kernel::Point_2 Cartesian_converter<Traits, Kernel>::operator()(Traits::Point_2)` must be available if `ExactAlphaComparisonTag` is set to `Tag_true`. That way we could event make it work for `Projection_traits` by adding a partial specialization to `Cartesian_converter`.

MaelRL replied:
Sorry, I had misunderstood. The new batch of commits goes in that direction (starting with  	db2556a).

afabri commented:
In Property_map it is a fix of a documentation.

afabri commented:
STL_extension is ok. It adds a policy similar to Location_policy.

janetournois commented:
I have a general comment.
In several files (including for example `Interpolation/include/CGAL/surface_neighbor_coordinates_3.h`, function `surface_neighbor_coordinates_certified_3`), the former code was
```
rt.insert(points_begin, points_end);
```
Now that implicit conversions have been removed, this code had to be changed by
```
for(Weighted_point_iterator wpit = points_begin; wpit != points_end; ++wpit)
  rt.insert(*wpit);
```
Should we add a function `insert(wp_begin, wp_end)` in `Regular_triangulation` ?
@MaelRL @MoniqueTeillaud

janetournois commented on Triangulation_3/doc/Triangulation_3/CGAL/Regular_triangulation_cell_base_3.h:
```diff
@@ -23,39 +23,57 @@ By default, this parameter is instantiated by
-typedef RegularTriangulationTraits_3::Bare_point Bare_point;
+typedef Traits::Point_3 Point_3;
+
+typedef Traits::Weighted_point_3 Point;
```
do we really want this `typedef`, where the weight of `Weighted_point_3` becomes somehow hidden?

janetournois commented on Triangulation_3/doc/Triangulation_3/CGAL/Regular_triangulation_cell_base_with_weighted_circumcenter_3.h:
```diff
@@ -13,58 +13,60 @@ Note that input/output operators discard this additional information.
-typedef RegularTriangulationTraits_3::Bare_point Bare_point;
+typedef Traits::Point_3 Point_3;
+
+typedef Traits::Weighted_point_3 Point;
```
same remark about the `typedef`

janetournois commented on Triangulation_3/doc/Triangulation_3/CGAL/Triangulation_3.h:
```diff
@@ -7,34 +7,36 @@ namespace CGAL {
-It has the default value `Triangulation_data_structure_3< Triangulation_vertex_base_3<TriangulationTraits_3>,Triangulation_cell_base_3<TriangulationTraits_3> >`. 
-`Default` may be used.
+\tparam TDS is the triangulation data structure and must be a model of `TriangulationDataStructure_3`.
+`Default` may be used, with default type is `Triangulation_data_structure_3<Triangulation_vertex_base_3<Traits>,
```
"with" should be "and", or "is" should be removed

janetournois commented on Triangulation_3/include/CGAL/Regular_triangulation_cell_base_3.h:
```diff
@@ -25,25 +25,31 @@
-
   typedef GT                                           Geom_traits;
+  typedef typename Geom_traits::Point_3                Point_3;
   typedef typename Geom_traits::Weighted_point_3       Point;
```
same remark about typedef

janetournois commented on Triangulation_3/include/CGAL/Regular_triangulation_cell_base_with_weighted_circumcenter_3.h:
```diff
@@ -26,24 +26,26 @@
-  typedef typename GT::Point_3                         Point_3;
-  typedef typename GT::Bare_point                      Bare_point;
+  typedef typename GT::Point_3                Point_3;
+  typedef typename GT::Weighted_point_3       Point;
```
same remark about typedef

janetournois commented on Installation/changes.html:
```diff
@@ -144,8 +161,83 @@ <h2 id="release4.11">Release 4.11 </h2>
+    <li><b>Breaking change</b>:
+      Uniformized the point type across all vertex and cell concepts. The triangulation
+      point type name is now always <code>Point</code>. Note that this does not change
+      the requirements but only the name: <code>Point</code> is stil expected to be equal to
```
"stil" should be "still"

janetournois commented on Installation/changes.html:
```diff
@@ -144,8 +161,83 @@ <h2 id="release4.11">Release 4.11 </h2>
+      point type name is now always <code>Point</code>. Note that this does not change
+      the requirements but only the name: <code>Point</code> is stil expected to be equal to
+      <code>Traits::Point_[23]</code> for basic and Delaunay triangulations or to
+      <code>Traits::Weighted_point_[23]</code> for regular triangulations. Consequently:
```
I guess this answers my question about typedefs in this package review. However I'm still wondering if it's the best option

MaelRL commented on Partition_2/include/CGAL/Partition_2/Triangulation_indirect_traits_2.h:
```diff
@@ -112,6 +112,13 @@ class Indirect_orientation_2
 
 template <class Circulator>
+class Construct_circulator_2
+{
```
[There](https://github.com/afabri/cgal/blob/67c531af5394e9faead89884a02cbc61e9b5c8c2/Partition_2/include/CGAL/Partition_2/Triangulation_indirect_traits_2.h#L146), just a consequence from requesting `Construct_point_2` in `TriangulationTraits_2`.

janetournois replied:
ok

MaelRL commented on Skin_surface_3/include/CGAL/Skin_surface_quadratic_surface_3.h:
```diff
@@ -151,44 +159,47 @@ class Skin_surface_quadratic_surface_3 {
-  FT Q[6]; 
-  Point p; 
+
+  int dim;
```
I moved it to be with the other members of the class, at the bottom (instead of being lost in the middle of functions).

janetournois replied:
ok!

sloriot commented:
I went quickly through kinetic things and it looks good.

sloriot commented on Alpha_shapes_2/doc/Alpha_shapes_2/CGAL/Alpha_shape_2.h:
```diff
@@ -16,20 +16,34 @@ and the \f$ k\f$-dimensional faces of the triangulation.
+
+\warning When the tag `ExactAlphaComparisonTag` is set to \link Tag_true `Tag_true`\endlink,
+the class `Cartesian_converter` is used internally to switch between the traits class
+and exact CGAL kernels. `Cartesian_converter` must thus provide the necessary functor
```
... and `CGAL::Simple_cartesian<NT>` where NT can be either `CGAL::Interval_nt` or `CGAL::Exact_rational`

MaelRL replied:
Changed in 41e7520.

sloriot commented on Alpha_shapes_3/include/CGAL/internal/Lazy_alpha_nt_3.h:
```diff
@@ -73,27 +105,26 @@ struct Types_for_alpha_nt_3< ::CGAL::Tag_true,Input_traits,Kernel_input,Kernel_a
   typedef CGAL::Simple_cartesian<NT_approx>                                                     Kernel_approx;
   typedef CGAL::Simple_cartesian<NT_exact>                                                      Kernel_exact;
+  typedef typename Kernel_traits<typename Input_traits::Point_3>::Kernel   Kernel_input;
+
```
With the documentation update, can't we simply skip the Kernel_traits and use the Traits directly in the converter?

MaelRL replied:
No, because the traits class might not derive from a kernel, and thus it might not know what e.g. a `Circle_3` is. Then, `Cartesian_converter<Traits, EK>` will not compile since it won't know how to compile `EK::Circle_3 operator()(const Traits::Circle_3)`.

janetournois commented:
From my side this PR is ok, apart from the discussion with @sloriot

MaelRL commented:
@lrineau  some CentOS test plateforms use boost 1.48, which does not have `function_property_map`. The pmap seems to appear somewhere around v1.51 (it's not in the changelog but copyright is 2012 and the library has a page from v1.51 on). 

I guess testing v1.48 is required for some reason, so what should be done?

Also, I had added my own version of this type of pmap initially into CGAL/Property_map.h  because I didn't know about boost's but I've removed it (see 99b77a54b703f913036767cad6a7fc09017149cb ) so I could use that, but it's redundant for everyone that has newer versions of boost.

afabri commented:
Note that instead of our own code, we could also include the one boos file, in case it han't evolved a lot.

sloriot commented:
Since you only have calls to `boost::make_function_property_map()` you can have `CGAL::make_function_property_map()` and use your implementation if boost is less than 1.48 and use boost otherwise, or always use yours (it's a really small code).

MaelRL commented:
Well, boost's version is slightly more generic than mine. Might as well put a copy of boost's version into `STL_Extension/include/CGAL/internal/boost`? That probably would be cleaner.

The code of `boost/property_map/function_property_map.hpp` hasn't changed since its introduction.

sloriot commented:
Sure! The funny thing is that @bo0ts is the author of this file :)

lrineau commented:
There are warnings on Windows that seem related to this branch:
https://cgal.geometryfactory.com/CGAL/testsuite/results-4.11-Ic-51.shtml#Kinetic_data_structures

lrineau commented:
As you mention #2212 depends on this PR #2102, I propose to merge #2102 in `master` anyway, and deal with the warnings of KDS in a later PR.

MaelRL commented:
Yes I've seen them. I propose to try and fix them with tonight's test suite and otherwise fix them later.

lrineau commented:
Alright. I will not merge this PR #2102 in `master` today.

bo0ts commented:
@sloriot I knew someone would find it useful some day! Thanks for pointing it out, good memories!

lrineau commented:
There are still [two warnings in KDS on a Windows platform (about ref to a temp)][w], but I propose to merge this PR in `master`, if you agree. That would unlock #2212.

  [w]: https://cgal.geometryfactory.com/CGAL/testsuite/results-4.11-Ic-53.shtml#Kinetic_data_structures

MaelRL commented:
I guess d68b8b6 should fix it, I assumed the mistake would only be in my new kernel in 9e16a94...

Can't reproduce locally so I didn't see it.

lrineau commented:
So we give it another day of testing?

MaelRL commented:
A possibility would be to merge, and I'll keep an eye on those two test suite blocks tomorrow and fix them in a separate PR or #2212 if d68b8b6 didn't fix it.

maxGimeno commented:
Since that PR was merged, the IO for binary c3t3 is broken. has Anyone any idea of what could have caused that ?
